Build WASM runtimes

vargaz opened the pull request.

vargaz committed:
Add beginnings of wasm binary generation.

vargaz commented:
The actual build is not setup yet.

marek-safar commented on src/mono/wasm/runtime/zlib-helper.c:
```diff
@@ -0,0 +1,373 @@
+/*
```
Why is it needed for net5?

vargaz replied:
Will remove the needless files later when doing the final sync from mono/mono.

marek-safar replied:
Why is it needed for net5?

marek-safar commented on src/mono/wasm/runtime/pinvoke-tables-default-netcore.h:
```diff
@@ -0,0 +1,229 @@
+// GENERATED FILE, DO NOT MODIFY
```
Do we have to commit generated file?

vargaz replied:
This file needs to be generated by a tool which is ran on the BCL assemblies. Not sure where this can be done during the build.

marek-safar replied:
It might be easier to run it after each native library build and include them as individual files.

@safern @ViktorHofer any recommendations?

safern replied:
How is this exactly generated? Do you run the tool against all managed assemblies?

vargaz replied:
This is now generated.

marek-safar commented on src/mono/wasm/runtime/pinvoke-tables-default.h:
```diff
@@ -0,0 +1,67 @@
+// GENERATED FILE, DO NOT MODIFY
```
Do we have to commit generated file?

vargaz commented on src/mono/wasm/Makefile:
```diff
@@ -0,0 +1,100 @@
+	cd $(EMSDK_PATH) && ./emsdk construct_env $(CURDIR)/emsdk_env.sh
+
+# artifacts/bin doesn't contain all the includes/libs we need
+ARTIFACTS_BIN_DIR=$(TOP)/artifacts/bin/mono/Browser.wasm.Release
+ARTIFACTS_OBJ_DIR=$(TOP)/artifacts/obj/mono/Browser.wasm.Release
+SYS_NATIVE_DIR=$(TOP)/artifacts/bin/native/netcoreapp5.0-Browser-Release-wasm/
```
Its not hooked up to msbuild yet.

vargaz commented on src/mono/wasm/Makefile:
```diff
@@ -0,0 +1,100 @@
+ARTIFACTS_BIN_DIR=$(TOP)/artifacts/bin/mono/Browser.wasm.Release
+ARTIFACTS_OBJ_DIR=$(TOP)/artifacts/obj/mono/Browser.wasm.Release
+SYS_NATIVE_DIR=$(TOP)/artifacts/bin/native/netcoreapp5.0-Browser-Release-wasm/
+MONO_LIBS = $(ARTIFACTS_OBJ_DIR)/out/lib/{libmono-ee-interp.a,libmonosgen-2.0.a,libmono-ilgen.a,libmono-icall-table.a} ${SYS_NATIVE_DIR}/libSystem.Native.a
+MONO_INCLUDE_DIR=$(ARTIFACTS_OBJ_DIR)/out/include/mono-2.0
```
Will do that once they contain these files.

vargaz committed:
Remove some unused files.

vargaz committed:
Hook the wasm build up to the build system.

The wasm build depends on libraries being built to compute pinvoke mapping tables, so it can only be
done during the installer build.

vargaz committed:
Add a WasmAppBuilder msbuild task. Use it to generate the pinvoke table.

vargaz commented on src/mono/wasm/runtime/linker-subs.xml:
```diff
@@ -0,0 +1,69 @@
+<linker>
```
Is it cleared up now, or it will be in the future ?

marek-safar replied:
You can consider now (basically imminent)

vargaz commented on eng/liveBuilds.targets:
```diff
@@ -137,6 +137,8 @@
       <MonoIncludeFiles Condition="'$(TargetsMobile)' == 'true' or '$(TargetOS)' == 'Browser'"
         Include="$(MonoArtifactsPath)\include\**\*.*" />
+      <WasmDistFiles Condition="'$(TargetOS)' == 'Browser'"
+        Include="$(MonoArtifactsPath)\wasm\runtimes\**\*.*" />
```
The different kinds of files go to different places inside the nuget, so they need different property names.

marek-safar replied:
I was referring to `\wasm\runtimes` not sure we need this if the package is wasm only

kjpou1 replied:
Nit:  Shouldn't we be using `Condition="'$(TargetsBrowser)' == 'true'"` here instead of `Condition="'$(TargetOS)' == 'Browser'"`

akoeplinger replied:
Ideally yes, but that can be cleaned up later.

vargaz committed:
Rewrite the pinvoke table generator to use System.Reflection.MetadataLoadContext instead of cecil.

vargaz committed:
Fix the build.

vargaz commented:
The wasm installer build needs EMSDK_PATH set as well.

marek-safar commented on src/mono/wasm/wasm.proj:
```diff
@@ -0,0 +1,37 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+	<WasmPInvokeModules Include="libSystem.Native"/>
```
What about all other native core libraries?

vargaz replied:
Added Sys.Console.

vargaz committed:
Remove linker subs.

vargaz committed:
Pass EMSDK_PATH to the wasm installer build.

vargaz committed:
Build WasmAppBuilder in wasm.proj if needed.

vargaz committed:
Publish WasmAppBuilder into a deterministic location.

vargaz committed:
Try running the wasm build in src/libraries/pretest.proj.

vargaz committed:
Use bash as the shell.

vargaz committed:
Get all runtime libs from bin/mono.

vargaz committed:
Add some debugging code.

vargaz committed:
Fix target ordering.

vargaz committed:
Use MonoArtifactsPath.

vargaz commented:
This seems to work now.

vargaz committed:
Remove debugging code.

marek-safar commented on src/mono/msbuild/WasmAppBuilder/WasmAppBuilder.cs:
```diff
@@ -0,0 +1,147 @@
+		var pinvokes = new List<PInvoke> ();
+
+		var resolver = new PathAssemblyResolver (assemblies);
+		var mlc = new MetadataLoadContext (resolver, "System.Private.CoreLib");
```
Why only SPC? Any src/library can include PInvoke?

vargaz replied:
Will add more libraries as needed. Running this is required to create the final executable, running it during the build would mean the executable needs to be recompiled/relinked, which takes a long time and requires emscripten.

vargaz replied:
wasm doesn't support dlopen, so these mapping tables need to be compiled into the executable. There are two modes right now:
- during development, we use a precompiled dotnet.wasm created here.
- for release, the same can be used, or its possible to recompile/relink it to make it smaller.

marek-safar replied:
Right, so we should run this for all 100 or so assemblies which are part of netcore as any of them can include pinvoke call.

marek-safar commented:
@akoeplinger please review

akoeplinger commented on src/mono/wasm/wasm.proj:
```diff
@@ -0,0 +1,43 @@
+      AssemblyFile="$(MonoProjectRoot)msbuild\WasmAppBuilder\bin\publish\WasmAppBuilder.dll"/>
+
+  <PropertyGroup>
+	<WasmPInvokeTablePath>$(MonoObjDir)wasm/pinvoke-table.h</WasmPInvokeTablePath>
```
fix the spaces vs. tabs mismatch here and in the rest of the file to only use spaces.

akoeplinger commented on src/mono/wasm/Makefile:
```diff
@@ -0,0 +1,105 @@
+EMCC=source $(CURDIR)/emsdk_env.sh && emcc
+
+emsdk:
+	git clone https://github.com/juj/emsdk.git $(EMSDK_PATH)
```
change this to the new url: https://github.com/emscripten-core/emsdk

akoeplinger commented on src/installer/pkg/projects/netcoreapp/src/netcoreapp.depproj:
```diff
@@ -71,11 +71,16 @@
 
+      <RuntimeFiles Condition="'$(TargetsBrowser)' == 'true'"
+        Include="@(WasmDistFiles)">
+        <TargetPath>runtimes/$(PackageRID)/native/wasm/runtimes/%(RecursiveDir)</TargetPath>
```
the `wasm/runtimes` in the path is redundant

vargaz committed:
Address comments.

vargaz committed:
Rename WasmAppBuilder task to PInvokeTableGenerator.

kjpou1 commented on src/mono/mono.proj:
```diff
@@ -982,6 +988,18 @@
+      </_MonoRuntimeArtifacts>
+      <_MonoRuntimeArtifacts Condition="'$(TargetOS)' == 'Browser'" Include="$(MonoObjDir)out\lib\libmono-profiler-aot.a">
+        <Destination>$(BinDir)libmono-profiler-aot.a</Destination>
+      </_MonoRuntimeArtifacts>
```
Nit: Shouldn't we be using Condition="'$(TargetsBrowser)' == 'true'" here instead of Condition="'$(TargetOS)' == 'Browser'" for consistency?

akoeplinger commented on src/mono/wasm/fix-emscripten-8511.diff:
```diff
@@ -0,0 +1,76 @@
+From d8a94684e2a9a584080757a324c4c42183f6d11a Mon Sep 17 00:00:00 2001
```
@vargaz this can be removed now that the patch was removed.

akoeplinger reviewed (approved):
I'm fine to merge to make progress, but please file an issue to track addressing the remaining cleanups/comments.

safern commented on src/mono/wasm/Directory.Build.props:
```diff
@@ -0,0 +1,3 @@
+<Project>
```
You don't need this file. If you don't include this file, `Directory.Build.props` in the above directory will be included by default.

vargaz replied:
Removing .props/.targets doesn't seem to work. According to the docs, its imported by
Microsoft.Common.props, do I need to import that instead ?

safern replied:
Oh, I thought this props where part of `WasmAppBuilder` project. The reason why this is not working on that project is because `wasm.proj` is not an SDK project:

https://github.com/dotnet/runtime/pull/35566/files/e7e0edc1ee2b97809844e2213574894674170d5f

You can achieve this by changing the `wasm.proj` to be an SDK project. I recommend using the Microsoft.Build.NoTargets SDK and that way you wouldn't need to have a `Directory.Build.props` nor `Directory.Build.targets` in this folder. You can do that by changing the project declaration in wasm.proj to:

`<Project Sdk="Microsoft.Build.NoTargets" DefaultTargets="Build">`

vargaz replied:
Done, thanks.

safern commented on src/mono/wasm/Directory.Build.targets:
```diff
@@ -0,0 +1,5 @@
+<Project>
```
Do we really need this? `Directory.Build.targets` is included recursively by MSBuild. So it will go up in the directory structure until it finds the first one and stop once it find it. I think if you remove this, you will get all the repo targets and the SDK targets as well.

vargaz committed:
Add more license headers.

vargaz committed:
Remove emscripten patch which is no longer needed.

vargaz committed:
Use TargetsBrowser in a few more places.

vargaz committed:
Remove zlib-helper.c, its not needed.

vargaz committed:
Convert wasm.proj to an SDK project.

safern commented on src/mono/wasm/wasm.proj:
```diff
@@ -1,6 +1,7 @@
-  <Import Project="Directory.Build.targets" />
+<Project Sdk="Microsoft.Build.NoTargets" DefaultTargets="Build">
+  <PropertyGroup>
+    <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
```
Did you really need to add a tfm for this?

vargaz replied:
I get:

NuGet.RestoreEx.targets(10,5): error : Invalid framework identifier ''. [/Users/vargaz/git/runtime/src/mono/wasm/wasm.proj]

otherwise.

safern replied:
It seems like somehow it is being restored, interesting. @ViktorHofer any suggestion here?

akoeplinger replied:
We're already doing the same thing in llvm.proj so this is fine.

safern commented on src/mono/wasm/wasm.proj:
```diff
@@ -35,9 +36,8 @@
   </Target>
 
-  <Target Name="BuildWasmRuntimes" DependsOnTargets="BuildPInvokeTable">
+  <Target Name="BuildWasmRuntimes" DependsOnTargets="BuildPInvokeTable" AfterTargets="Build">
```
I would just make this target the default target instead of making it `AfterTargets="Build"`

vargaz committed:
Make BuildWasmRuntimes the default target.

vargaz committed:
Fix the build.

vargaz committed:
use TargetOS==Browser in pretest.proj, TargetsBrowser is not defined there.

vargaz commented:
Rebased.

akoeplinger commented on src/mono/wasm/Makefile:
```diff
@@ -0,0 +1,104 @@
+	cd $(EMSDK_PATH) && ./emsdk construct_env $(CURDIR)/emsdk_env.sh
+
+MONO_OBJ_DIR=$(OBJDIR)/mono/Browser.wasm.$(CONFIG)
+SYS_NATIVE_DIR=$(BINDIR)/native/netcoreapp5.0-Browser-$(CONFIG)-wasm/
```
@vargaz this should fix the current build failure:

```suggestion
SYS_NATIVE_DIR=$(BINDIR)/native/net5.0-Browser-$(CONFIG)-wasm/
```

Ideally we'd not deal with this in Makefile...

vargaz committed:
Fix the build.

vargaz committed:
Add System.Console to the pinvoke table generator. Add pinvoke-table.h to the make dependencies.

vargaz committed:
Simplify local emsdk installation.

vargaz committed:
Fix the build.

vargaz committed:
Load libs from a different path for the pinvoke task, lib-runtime-packs might not exist yet.

vargaz commented:
This seems to work now.

akoeplinger commented on src/mono/msbuild/WasmAppBuilder/WasmAppBuilder.csproj:
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <OutputType>Library</OutputType>
+    <OutputPath>bin</OutputPath>
+    <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
```
There was a recent change to move the AppBuilders to artifacts/ in https://github.com/dotnet/runtime/commit/741e9054d25adc8a93fd2695b0848adcecaaffb0 so you can remove this:

```suggestion
    <OutputPath>bin</OutputPath>
    <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
```

You'll need to change the paths where it is referenced.

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -945,9 +944,16 @@
+  <!-- Need to publish this since it references System.Reflection.MetadataLoadContext -->
+  <Target Name="BuildWasmAppBuilder">
+    <MSBuild Projects="$(MonoProjectRoot)msbuild\WasmAppBuilder\WasmAppBuilder.csproj"
+             Properties="Configuration=$(Configuration);OutputPath=$(MonoProjectRoot)msbuild\WasmAppBuilder\bin\"
```
OutputPath is no longer needed:

```suggestion
             Properties="Configuration=$(Configuration)"
```

akoeplinger commented on src/mono/wasm/wasm.proj:
```diff
@@ -0,0 +1,46 @@
+  </PropertyGroup>
+
+  <UsingTask TaskName="PInvokeTableGenerator"
+      AssemblyFile="$(MonoProjectRoot)msbuild\WasmAppBuilder\bin\publish\WasmAppBuilder.dll"/>
```
```suggestion
      AssemblyFile="$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'WasmAppBuilder', '$(NetCoreAppCurrent)-$(MonoConfiguration)'))publish\WasmAppBuilder.dll"/>
```

vargaz committed:
Load WasmAppBuilder from artifacts/bin/WasmAppBuilder.

vargaz committed:
Put wasm files into lib-runtime-packs.

vargaz committed:
Fix nullability warnings.
